docs/01-Protocol-Versions/Version6.md

```diff
@@ -0,0 +1,198 @@
+      values at this step (e.g. `v6c.local.`).
+4. Decode the payload (`m` sans `h`, `f`, and the optional trailing period
+   between `m` and `f`) from base64url to raw binary. Set:
+    * `n` to the leftmost 32 bytes
+    * `t` to the rightmost 32 bytes
+    * `c` to the middle remainder of the payload, excluding `n` and `t`.
+5. Split the key into an Encryption key (`Ek`) and Authentication key (`Ak`),
+   using keyed BLAKE2b, using the domain separation constants and `n` as the
+   message, and the input key as the key. The first value will be 56 bytes,
+   the second will be 32 bytes.
+   The derived key will be the leftmost 32 bytes of the hash output.
+   The remaining 24 bytes will be used as a counter nonce (`n2`):
+   ```
+   tmp = crypto_generichash(
+       msg = "paseto-encryption-key" || n,
+       key = key,
+       length = 56
+   );
+   Ek = tmp[0:32]
+   n2 = tmp[32:]
+   Ak = crypto_generichash(
+       msg = "paseto-auth-key-for-aead" || n,
+       key = key,
+       length = 32
+   );
+   ```
+6. Pack `h`, `n`, `c`, `f`, and `i` together (in that order) using
+   [PAE](Common.md#authentication-padding).
+   We'll call this `preAuth`.
+7. Re-calculate BLAKE2b-MAC of the output of `preAuth`, using `Ak` as the
+   authentication key. We'll call this `t2`.
+   ```
+   t2 = crypto_generichash(
+       message = preAuth
+       key = Ak,
+       length = 32
+   );
+   ```
+8. Compare `t` with `t2` using a constant-time string compare function. If they
+   are not identical, throw an exception.
+    * You **MUST** use a constant-time string compare function to be compliant.
+      If you do not have one available to you in your programming language/framework,
+      you MUST use [Double HMAC](https://paragonie.com/blog/2015/11/preventing-timing-attacks-on-string-comparison-with-double-hmac-strategy).
+9. Decrypt `c` using `XChaCha20`, store the result in `p`.
+   ```
+   p = crypto_stream_xchacha20_xor(
+      ciphertext = c
+      nonce = n2
+      key = Ek
+   );
+   ```
+10. If decryption failed, throw an exception. Otherwise, return `p`.
+
+## Sign
+
+Given a message `m`, 64-byte SLH-DSA-SHA256-128s secret key `sk`  an optional footer `f`
+(which defaults to empty string), and an optional implicit assertion `i` (which
+defaults to empty string):
+
+1. Before signing, first assert that the key being used is intended for use
+   with `v6.public` tokens, and is the secret key of the intended keypair.
+   See [Algorithm Lucidity](../02-Implementation-Guide/03-Algorithm-Lucidity.md)
+   for more information.
+2. Set `h` to `v6.public.`  
+   **Note**: This includes the trailing period.
+3. Pack `pk`, `h`, `m`, `f`, and `i` together using
+   [PAE](Common.md#authentication-padding)
+   (pre-authentication encoding). We'll call this `m2`.
+    * Note: `pk` is the public key corresponding to `sk`. `pk` **MUST** be
+      32 bytes long.
+4. Sign `m2` using SLH-DSA-SHA256-128s with the private key `sk`. We'll call this `sig`.  
```

Comment:
(sorry if this was supposed to be a separate issue re #37, I can close and open as issue)

Would this be signing deterministically given `sk` or does it imply randomized signatures? Apologies if I missed something obvious.

Reply:
From [FIPS 205](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.205.pdf), section 9.2:

> The message randomizer may be set in either a non-deterministic or deterministic way, depending
on whether 𝑎𝑑𝑑𝑟𝑛𝑑 is provided as an input. For the “hedged” variant, 𝑎𝑑𝑑𝑟𝑛𝑑 is provided as an
input, and 𝑜𝑝𝑡_𝑟𝑎𝑛𝑑 is set to 𝑎𝑑𝑑𝑟𝑛𝑑. The hedged variant is the default and should be used on
platforms where side-channel attacks are a concern. When using the hedged version, 𝑎𝑑𝑑𝑟𝑛𝑑
shall be an 𝑛-byte random value. While 𝑎𝑑𝑑𝑟𝑛𝑑 should ideally be generated by an approved
random bit generator, other methods for generating fresh random values may be used.